### PR TITLE
filters/auth: add host opt-out to jwtMetrics and oauthTokeninfoValidate

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Extend configuration of `jwtMetrics` and `oauthTokeninfoValidate` to support opt-out by request host pattern - disable metrics collection and validation when request host matches any of the configured opt-out regular expressions.

This can be used to exclude internal cluster domain (*.ingress.cluster.local hosts).

### CODE REVIEW

File: filters/auth/jwt_metrics_test.go [hunk 1]
```diff
@@ -223,7 +225,44 @@ func TestJwtMetrics(t *testing.T) {
...
+			status: http.StatusOK,
+			expected: map[string]int64{
+				"jwtMetrics.custom.GET.foo_test.200.invalid-token": 1,
+			},
+			expectedTag: "invalid-token",
+		},
+		{
+			name:     "no metrics when host matches opted-out domain",
+			filters:  `jwtMetrics("{issuers: [foo, bar], optOutHosts: [ '^.+[.]domain[.]test$', '^exact[.]test$' ]}")`,
+			request:  &http.Request{Method: "GET", Host: "foo.domain.test"},
```
Comment (RomanZavodskikh) [inline filters/auth/jwt_metrics_test.go:240]: Should we test the same filter, but with "domain.test" host?

File: filters/auth/jwt_metrics_test.go [hunk 2]
```diff
@@ -223,7 +225,44 @@ func TestJwtMetrics(t *testing.T) {
...
+			name:     "no metrics when second level host matches opted-out domain",
+			filters:  `jwtMetrics("{issuers: [foo, bar], optOutHosts: [ '^.+[.]domain[.]test$', '^exact[.]test$' ]}")`,
+			request:  &http.Request{Method: "GET", Host: "foo.bar.domain.test"},
+			status:   http.StatusOK,
+			expected: map[string]int64{},
+		},
+		{
+			name:     "no metrics when host matches opted-out host exactly",
+			filters:  `jwtMetrics("{issuers: [foo, bar], optOutHosts: [ '^.+[.]domain[.]test$', '^exact[.]test$' ]}")`,
+			request:  &http.Request{Method: "GET", Host: "exact.test"},
```
Comment (RomanZavodskikh) [inline filters/auth/jwt_metrics_test.go:254]: Should we test the same filter, but with "not.exact.test" and/or "notexact.test" host?

File: filters/auth/tokeninfo_test.go [hunk 3]
```diff
@@ -669,16 +676,35 @@ func TestOAuthTokeninfoValidate(t *testing.T) {
...
+			expectStatus:       http.StatusOK,
+			expectAuthRequests: 0,
+		},
+		{
+			name:               "allow missing token when request host matches opted-out host exactly",
+			host:               "exact.test",
+			authHeader:         "",
+			expectStatus:       http.StatusOK,
+			expectAuthRequests: 0,
+		},
```
Comment (RomanZavodskikh) [inline filters/auth/tokeninfo_test.go:699]: Should we test "domain.test" and "not.exact.test" and/or "notexact.test" here as well?

File: filters/auth/tokeninfo_test.go [hunk 4]
```diff
@@ -713,3 +743,45 @@ func TestOAuthTokeninfoValidate(t *testing.T) {
...
+	tio := TokeninfoOptions{
+		URL:     "https://auth.test",
+		Timeout: testAuthTimeout,
+	}
+	spec := NewOAuthTokeninfoValidate(tio)
+
+	t.Run("valid", func(t *testing.T) {
+		for _, def := range []string{
+			`oauthTokeninfoValidate("{}")`,
+			`oauthTokeninfoValidate("{optOutAnnotations: [oauth.disabled], optOutHosts: [ '^.+[.]domain[.]test$', '^exact[.]test$' ], unauthorizedResponse: 'Authentication required, see https://auth.test/foo'}")`,
```
Comment (RomanZavodskikh) [inline filters/auth/tokeninfo_test.go:757]: I would've added oauthTokeninfoValidate fiters with single arguments as well, like this:
+ `oauthTokeninfoValidate("{optOutAnnotations: [oauth.disabled], optOutHosts: [ '^.+[.]domain[.]test$', '^exact[.]test$' ]}")`
+ `oauthTokeninfoValidate("{unauthorizedResponse: 'Authentication required, see https://auth.test/foo'}")`

File: filters/auth/jwt_metrics_test.go (hunk 2 above)
Comment (AlexanderYastrebov) [reply filters/auth/jwt_metrics_test.go:254]: There is "counts invalid-token when host does not match" test case for that.

File: filters/auth/jwt_metrics_test.go (hunk 1 above)
Comment (AlexanderYastrebov) [reply filters/auth/jwt_metrics_test.go:240]: We can but the goal is not to test regexp itself or all edge cases but the logic when host matches any or none of the configured patterns.

File: filters/auth/tokeninfo_test.go (hunk 4 above)
Comment (AlexanderYastrebov) [reply filters/auth/tokeninfo_test.go:759]: This quickly goes out of hands but I added.

**Comment by RomanZavodskikh** [comment]
:+1:

File: filters/auth/tokeninfo_test.go [hunk 5]
```diff
@@ -713,3 +743,47 @@ func TestOAuthTokeninfoValidate(t *testing.T) {
...
+			`oauthTokeninfoValidate(1)`,
+			`oauthTokeninfoValidate("{optOutAnnotations: [oauth.disabled]}", "extra arg")`,
+			`oauthTokeninfoValidate("{optOutHosts: [ '[' ]}")`, // invalid regexp
+		} {
+			t.Run(def, func(t *testing.T) {
+				f := eskip.MustParseFilters(def)[0]
+				require.Equal(t, spec.Name(), f.Name)
+
+				_, err := spec.CreateFilter(f.Args)
+				t.Logf("%v", err)
```
Comment (szuecs) [inline filters/auth/tokeninfo_test.go:784]: does not assert.Error show the same thing?

Comment (AlexanderYastrebov) [reply filters/auth/tokeninfo_test.go:784]: I think assert.Error does not log the error, only checks that is not nil
